### ISSUE TITLE
Generating str/int/float/bool values with GPT2 using generate.json leads to absurd results

### ISSUE THREAD

**Issue by rlouf** [outlines-dev/outlines#472]
### Discussed in https://github.com/outlines-dev/outlines/discussions/461

<div type='discussions-op-text'>

<sup>Originally posted by **gaspard-dv** December 19, 2023</sup>
### Describe the issue as clearly as possible:

I experimented a bit with outlines and I wonder what would be the recommended way to generate bools/numbers/ints.

* When I use `generate.format(model, <type>, allow_empty_tokens=False)` the resulting response is often just an empty string, which can't be converted to the desired `<type>`.
* When I use `generate.json(model, json.dumps({"type": "<type_in_json_schema_style>"})` the model tends to "block" or to produce stupid answers.

Regardless of my question, I believe that this second bullet point is an issue, so I'm adding a reproduction snippet below.

*NB: I'm not allowed to add the "correctness" label.*

### Steps/code to reproduce the bug:

```python
from json import dumps

from outlines import models
from outlines.text.generate import continuation
from outlines.text.generate import json
import torch

model = models.transformers("gpt2")
prompt = "a b c d e f"

rng = torch.Generator()

rng.manual_seed(789001)
print(continuation(model, max_tokens=30)(prompt, rng=rng))

rng.manual_seed(789001)
print(json(model, dumps({"type": "string"}))(prompt, rng=rng))
```

### Expected result:

```shell
EXPECTED:
g h i j k (...)
g h i j k (...)

ACTUAL:
g h i j k (...)
for all readable words H. Servo Oamber Trident Warrior Drug of others (...)
```

### Outlines/Python version information:

```python
Outlines 0.0.13
Python 3.10.12
```

### Context for the issue:

_No response_</div>

**Comment by RobinPicard** [comment]
I think it may be because the regex corresponding to a string type in the json schema includes double quotes at the start and at the end (`STRING = f'"{STRING_INNER}*"'`)
The `regex_pattern` created_by `interegular.parse_pattern(regex_string)` for the example given above with `json` is:
```
Pattern:
  Concatenation:
    _CharGroup(chars=frozenset({'"'}), negated=False)
    Repeated[0:]:
        Pattern:
          Concatenation:
            _CharGroup(chars=frozenset({'\x93', '\x8e', '\x10', '\x9c', '\x15', '\x8a', '\x88', '\x03', '\x1b', '\t', '\\', '\x86', '\x96', '\x00', '"', '\x97', '\x92', '\x0f', '\x04', '\x89', '\x02', '\x18', '\x19', '\x11', '\x14', '\x98', '\x87', '\x91', '\x17', '\x99', '\x7f', '\x81', '\x1a', '\x85', '\x01', '\x0b', '\x06', '\x0c', '\x84', '\x1c', '\x80', '\x8f', '\x83', '\x94', '\x9d', '\x90', '\x95', '\x08', '\x05', '\n', '\r', '\x12', '\x1f', '\x8c', '\x9a', '\x0e', '\x9b', '\x9f', '\x8b', '\x9e', '\x8d', '\x16', '\x82', '\x1e', '\x07', '\x13', '\x1d'}), negated=True)
          Concatenation:
            _CharGroup(chars=frozenset({'\\'}), negated=False)
            __DotCls()
    _CharGroup(chars=frozenset({'"'}), negated=False)
```
If you use the same example but remove the 2 double quotes from the `STRING` definition, then you get `g` as an output.

**Comment by rlouf** [comment]
In this case I would say this is expected behavior, and not a bug. At most it should be documented somewhere.

**Comment by RobinPicard** [comment]
Maybe we could add string as one of the supported python types of the `format` function such that people would use it instead of trying to tweak the `json` one as in the example above

**Comment by rlouf** [comment]
It's a bit redundant with `generate.text` but why not
